[WebDriver][WPE][GTK] Enable WebDriver connection to already running browser

zhani opened the pull request:
#### 56dbdddc059f3ee1689f34d8626600722aebcc73
<pre>
[WebDriver][WPE][GTK] Enable WebDriver connection to already running browser
<a href="https://bugs.webkit.org/show_bug.cgi?id=270630">https://bugs.webkit.org/show_bug.cgi?id=270630</a>

Reviewed by Carlos Garcia Campos.

In some embedded cases WebDriver needs to be able to connect to already
existing instance of browser running automation session. There&apos;s already
code in WebDriver to support this scenario but it&apos;s not used in WPE/GTK.

* Source/WebDriver/Capabilities.h:
* Source/WebDriver/SessionHost.h:
* Source/WebDriver/WebDriverService.cpp:
(WebDriver::printUsageStatement):
(WebDriver::WebDriverService::run):
(WebDriver::WebDriverService::connectToBrowser):
* Source/WebDriver/WebDriverService.h:
* Source/WebDriver/glib/SessionHostGlib.cpp:
(WebDriver::SessionHost::isConnected const):
(WebDriver::SessionHost::launchBrowser):
(WebDriver::SessionHost::connectToBrowser):
(WebDriver::SessionHost::connectionDidClose):
* Source/WebDriver/gtk/WebDriverServiceGtk.cpp:
(WebDriver::WebDriverService::platformParseCapabilities const):
* Source/WebDriver/wpe/WebDriverServiceWPE.cpp:
(WebDriver::WebDriverService::platformParseCapabilities const):

Canonical link: <a href="https://commits.webkit.org/276231@main">https://commits.webkit.org/276231@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/aad7cf914174bfb2fb1c939c8d890c195ff11ad3

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/43907 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/22961 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/46337 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/46551 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/39982 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/26852 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/20359 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/36227 "Failure limit exceed. At least found 2 new test failures: compositing/animation/animation-compositing.html, compositing/animation/busy-indicator.html (failure)") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/44482 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/19983 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/37813 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/17230 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/17500 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/38898 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/1963 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/40070 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/39193 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/48116 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/18911 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/15486 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/43044 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/20302 "Built successfully") | | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/41762 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/20506 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/6049 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/19926 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

carlosgcampos commented on Source/WebDriver/Capabilities.h:
```diff
@@ -83,7 +83,7 @@ struct Capabilities {
     std::optional<bool> useOverlayScrollbars;
 #endif
-#if USE(INSPECTOR_SOCKET_SERVER)
+#if USE(INSPECTOR_SOCKET_SERVER) || PLATFORM(GTK) || PLATFORM(WPE)
```
This is now all the users of Source/WebDriver, so you can just remove the ifdefs

zhani replied:
Good point. Fixed

carlosgcampos commented on Source/WebDriver/SessionHost.cpp:
```diff
@@ -27,7 +27,6 @@
 #include "SessionHost.h"
 
 #include <wtf/text/StringBuilder.h>
-
```
Don't remove this line

zhani replied:
Fixed

carlosgcampos commented on Source/WebDriver/SessionHost.h:
```diff
@@ -122,11 +122,18 @@ class SessionHost
+
+#if PLATFORM(GTK) || PLATFORM(WPE)
+    bool m_isRemoteBrowser { false };
+#endif
```
This could be in the USE(GLIB) block

zhani replied:
Done

carlosgcampos commented on Source/WebDriver/WebDriverService.cpp:
```diff
@@ -50,16 +50,16 @@ static void printUsageStatement(const char* programName)
-#if USE(INSPECTOR_SOCKET_SERVER)
-    printf("  -t <ip:port> --target=<ip:port> [WinCairo] Target IP and port\n");
+#if USE(INSPECTOR_SOCKET_SERVER) || PLATFORM(GTK) || PLATFORM(WPE)
+    printf("  -t <ip:port> --target=<ip:port> [GTK][WinCairo][WPE] Target IP and port\n");
```
We can just remove the port names

zhani replied:
Done

carlosgcampos commented on WebDriverTests/imported/selenium/py/conftest.py:
```diff
@@ -49,6 +49,10 @@ def pytest_addoption(parser):
+    parser.addoption('--browser-target-ip', action='store', dest='browser_ip',
+                     help='ip of exiting browser instance running in automation mode')
+    parser.addoption('--browser-target-port', action='store', dest='browser_port',
+                     help='port of exiting browser instance running in automation mode')
```
We should not change the imported selenium, except when updating from upstream. Are these changes upstream already? or proposed?

zhani replied:
These options are mandatory in order to tests to work. Without these we cannot connect to running remote instance and if these are not present and we still pass them from test suite executor then test fails when trying to parse non supported options.
Some of the changes in imported selenium are obsolete if we take upstream but for example these options need to be added. I'm going to propose all of the needed changes to upstream. We could wait for upstream PRs to land and then update imported scripts but not sure how long will it take. Also without these parameters we cannot land wk test suite script changes in this PR

carlosgcampos replied:
I'm not saying we don't needs those options, I was wondering if those are changes already un selenium or not. Because the policy is to not change imported code except when updating from upstream. So yes, I think we should first make sure upstream selenium has what we need, then update imported selenium and then use it from WebKit.

zhani replied:
ok let's try pushing changes in imported folder to selenium upstream first

lauromoura replied:
Regarding the `Tools/` and `WebDriverTests/` changes, what if we remove them, leaving the patch only adding the options to the WebDriver service? As this feature is not exercised by the bots, we can still test it manually, unblocking the patch while the Selenium part is sorted out.

Once we update `WebDriverTests/imported/selenium`, we can add the `Tools/` changes adding the new options back in a separate PR.

carlosgcampos replied:
Sure, it's ok to land the reest of the patch

zhani replied:
Ok, patch is now updated and Tools and WebDriverTests changes are now removed. Will push them later when we get support to selenium

carlosgcampos commented on Source/WebDriver/SessionHost.h:
```diff
@@ -116,14 +114,16 @@ class SessionHost
     GRefPtr<GSubprocess> m_browser;
     RefPtr<SocketConnection> m_socketConnection;
     GRefPtr<GCancellable> m_cancellable;
+    bool m_isRemoteBrowser { false };
```
What does remote means here? I guess this means it's an existing browser instance, not spawned by the driver, right? I think `m_isExistingBrowser` would be clearer name.

zhani replied:
Remote means that browser instance is not hosted by WebDriver and not using localhost address 127.0.0.1 for inspector. So in that sense it's remote browser from webdriver point of view.

carlosgcampos replied:
Existing browser is clearer to me, but feel free to leave remote browser if you prefer

zhani replied:
I would like to leave it as m_isRemoteBrowser

carlosgcampos commented on Source/WebDriver/glib/SessionHostGlib.cpp:
```diff
@@ -134,43 +134,58 @@ static guint16 freePort()
+    GUniquePtr<char> inspectorAddress(
+        g_strdup_printf("%s:%u", targetIp.isEmpty()?"127.0.0.1":targetIp.latin1().data(), targetPort > 0 ? targetPort :freePort())
+    );
+    if (targetIp.isEmpty()) {
```
We can return early here instead and avoid all the re-indentation.

zhani replied:
There's connectToBrowser call at the end of the function that still needs to be called in both cases

carlosgcampos replied:
Yes, just include the call before the return.

zhani replied:
Fixed

carlosgcampos commented on Source/WebDriver/glib/SessionHostGlib.cpp:
```diff
@@ -134,10 +134,28 @@ static guint16 freePort()
+
     m_cancellable = adoptGRef(g_cancellable_new());
+    GUniquePtr<char> inspectorAddress(
+        g_strdup_printf("%s:%u", targetIp.isEmpty()?"127.0.0.1":targetIp.latin1().data(), targetPort > 0 ? targetPort :freePort())
```
```suggestion
        g_strdup_printf("%s:%u", targetIp.isEmpty() ? "127.0.0.1" : targetIp.latin1().data(), targetPort > 0 ? targetPort : freePort())
```

zhani replied:
Fixed

zhani committed:
[WebDriver][WPE][GTK] Enable WebDriver connection to already running browser
https://bugs.webkit.org/show_bug.cgi?id=270630

Reviewed by Carlos Garcia Campos.

In some embedded cases WebDriver needs to be able to connect to already
existing instance of browser running automation session. There's already
code in WebDriver to support this scenario but it's not used in WPE/GTK.

* Source/WebDriver/Capabilities.h:
* Source/WebDriver/SessionHost.h:
* Source/WebDriver/WebDriverService.cpp:
(WebDriver::printUsageStatement):
(WebDriver::WebDriverService::run):
(WebDriver::WebDriverService::connectToBrowser):
* Source/WebDriver/WebDriverService.h:
* Source/WebDriver/glib/SessionHostGlib.cpp:
(WebDriver::SessionHost::isConnected const):
(WebDriver::SessionHost::launchBrowser):
(WebDriver::SessionHost::connectToBrowser):
(WebDriver::SessionHost::connectionDidClose):
* Source/WebDriver/gtk/WebDriverServiceGtk.cpp:
(WebDriver::WebDriverService::platformParseCapabilities const):
* Source/WebDriver/wpe/WebDriverServiceWPE.cpp:
(WebDriver::WebDriverService::platformParseCapabilities const):

Canonical link: https://commits.webkit.org/276231@main
